### PR TITLE
feat(snql-sdk) Add a flag to use the SnQL SDK to run a snuba query

### PR DESCRIPTION
**Pull request by evanh**
SnQL and the SDK are still in the testing phase, so turning on the flag doesn't
actually submit the query using SnQL. Instead for now, it will run the query as
normal using the legacy endpoint, and asynchronously call the SnQL endpoint as a
dry run. If the generated SQL from the two calls is not the same, it logs an
error. That way we can turn the flag on, ensure there are no errors before we
flip over to the SnQL endpoint.

Dry run calls do not impact rate limits at all in Snuba, so the extra query only
adds a little bit of overhead.

### CODE REVIEW

File: src/sentry/utils/snuba.py [hunk 1]
```diff
@@ -764,6 +790,100 @@ def _snuba_query(params):
...
+        query.validate()  # Call this here just avoid it happening in the async all
+    except Exception as e:
+        logger.warning(
+            "snuba.snql.parsing.error",
+            extra={"error": str(e), "params": json.dumps(query_params)},
+        )
+        return _snuba_query(params)
+
+    query = query.set_dry_run(True).set_debug(True)
+    query_params["debug"] = True
```
Comment (wmak) [inline src/sentry/utils/snuba.py:811]: I've always been curious, but do we know if adding debug changes the perf of the query significantly?

Comment (evanh) [reply src/sentry/utils/snuba.py:811]: I can look into it. It shouldn't at all, except that maybe the network IO is slightly higher.

**Review by wmak — APPROVED** [review]
:+1: Other than being curious about debug perf, this change LGTM.

When we start enabling it are we going to do it slowly across queries with referrer? Or is there another strategy in mind?

**Comment by evanh** [comment]
> When we start enabling it are we going to do it slowly across queries with referrer? Or is there another strategy in mind?

I'm going to add a feature flag, and then add that feature flag to some different endpoint(s) and slowly ramp up the percentage. Once I'm confident things are working I can migrate over more and more.

**Comment by fpacifici** [comment]
Please ensure that we add the SnQL query in parallel slowly and that we do not duplicate all queries. Even if the SnQL queries do not impact the rate limit and do not take resources from Clickhouse, they still run through the entire query processing so take cSnuba cpu and snuba threads. Most of snuba load is on very short query where the query processing phase take 40% of the query execution time. Doubling that would be a problem.

File: src/sentry/utils/snuba.py [hunk 2]
```diff
@@ -764,6 +790,100 @@ def _snuba_query(params):
...
+
+    legacy_future = _query_thread_pool.submit(_snuba_query, (query_data, Hub(thread_hub), headers))
+    snql_future = _query_thread_pool.submit(_raw_snql_query, query, Hub(thread_hub), headers)
```
Comment (wedamija) [inline src/sentry/utils/snuba.py:814]: Does it make sense at all to just submit `snql_future` to the thread pool then call `_snuba_query` directly?

**Review by wedamija — APPROVED** [review]

File: src/sentry/utils/snuba.py (hunk 2 above)
Comment (evanh) [reply src/sentry/utils/snuba.py:814]: Yeah that's a little cleaner I think.

**Pull request merged**